### PR TITLE
Refactor Android <-> JS storage layer

### PR DESCRIPTION
**Pull request by thisandagain**
Resolves GH-2007
Unblocks GH-1670

### CODE REVIEW

**Comment by thisandagain** [comment]
@gvn and @Pomax R?

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java [hunk 1]
```diff
@@ -118,7 +129,7 @@ public void run() {
     public void setView(final String url, final String routeData) {
         Activity activity = (Activity) mContext;
         if (activity == null) return;
-        MemStorage.sharedCache().put(url, routeData);
+        MemStorage.sharedStorage().put("route::data", routeData);
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: let's pack that string up as a constant or enum

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java [hunk 2]
```diff
@@ -128,35 +139,20 @@ public void run() {
...
     }
 
     @JavascriptInterface
-    public void getFromMedia() {
-        Element elementActivity = (Element) mContext;
-        if (elementActivity != null) {
-            elementActivity.dispatchMediaIntent();
-        }
+    public String getRouteData() {
+        return MemStorage.sharedStorage().get("route::data");
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: magic string instance #2. Definitely needs packing up.

File: app/src/main/java/mozilla/org/webmaker/storage/MemStorage.java [hunk 3]
```diff
@@ -1,61 +1,20 @@
...
 
-import android.app.ActivityManager;
-import android.content.Context;
 import android.util.LruCache;
 
 public class MemStorage {
 
-    private static Context context;
-    private static int cacheSize = 5 * 1024 * 2014;
+    private static int cacheSize = 24 * 1024 * 2014;
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/storage/MemStorage.java]: what's this new number based on? (we might want to document that in the code so we know what it should reflect and whether it's gone stale)

File: www_src/lib/api.js [hunk 4]
```diff
@@ -54,11 +54,11 @@ module.exports = function (options, callback) {
...
     // Set cache if window.Android is available
     if (window.Android) {
-      window.Android.setSharedPreferences(key, JSON.stringify(body), false);
+      window.Android.setMemStorage(key, JSON.stringify(body), false);
     }
 
     // If there is a callback, forward the response body
-    if(callback) {
+    if (callback) {
       callback(null, body);
```
Comment (Pomax) [inline www_src/lib/api.js]: nit: there is no good reason to ever use `null` in JS context. Can we make this "false" if we're editing this file anyway?

File: www_src/lib/router.jsx [hunk 5]
```diff
@@ -30,9 +30,22 @@ module.exports = {
 
     return r;
   },
+  getRouteData: function () {
+    var r = {};
```
Comment (Pomax) [inline www_src/lib/router.jsx]: nit: minification will take care of `r`, let's call it `routes` or something in the full code?

**Comment by Pomax** [comment]
question: since the vast majority of the time memCache is going to be scoped, does it make sense to have a "scope" boolean, rather than a "global" boolean? And then have memCache be automatically scoped unless explicitly unscoped?

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java [hunk 6]
```diff
@@ -33,59 +33,76 @@ public WebAppInterface(Context context, JSONObject routeParams) {
...
     /**
      * ---------------------------------------
-     * SharedPreferences
+     * Disk-based Storage
      * ---------------------------------------
      */
 
     @JavascriptInterface
-    public String getSharedPreferences(String key, boolean scope) {
+    public String getSharedPreferences(String key, final boolean scope) {
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: good idea to add a functor overload with a default, like
```
public String getSharedPreferences(String key) {
  return getSharedPreferences(key, true);
}
```

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java [hunk 7]
```diff
@@ -33,59 +33,76 @@ public WebAppInterface(Context context, JSONObject routeParams) {
...
-        if (scope) {
-            key = key.concat(mPrefKey);
-        }
+        if (scope) key = key.concat(mPrefKey);
         return getter.getString(key, null);
     }
 
     @JavascriptInterface
-    public void setSharedPreferences(String key, String value, boolean scope) {
+    public void setSharedPreferences(String key, final String value, final boolean scope) {
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: same overloading note

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java [hunk 8]
```diff
@@ -33,59 +33,76 @@ public WebAppInterface(Context context, JSONObject routeParams) {
...
+    /**
+     * ---------------------------------------
+     * Memory-based Storage
+     * ---------------------------------------
+     */
+
     @JavascriptInterface
-    public String getState() {
-        return mPrefs.getString("page_state", "{}");
+    public String getMemStorage (String key, final boolean scope) {
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: ... etc =)

**Comment by Pomax** [comment]
Runtime-wise I don't see anything blowing up, although the proliferation of makes (due to everyone making with the same user account) is definitely interfering with detailed testing.

**Comment by thisandagain** [comment]
Thanks for the great review @Pomax . Feedback addressed in the latest commit.

Re-scope vs. global. It's fairly split right now and my guess is that we'll increase the sharing of objects between views (scope = false) over time as we optimize performance – thus the initial "scope" naming.

**Comment by thisandagain** [comment]
That said, I think "global" is easier to understand. Will revise.

**Comment by gvn** [comment]
Looks good. It would be nice to have a browser polyfill for dev outside of Android, but I'd say a follow up ticket for that is fine since it's a nice-to-have for developer ergonomics.

R+!

dance2

**Pull request merged**